service/share/discovery.go

```diff
@@ -0,0 +1,134 @@
+package share
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	core "github.com/libp2p/go-libp2p-core/discovery"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p/p2p/discovery/util"
+)
+
+const (
+	// PeersLimit is max amount of peers that will be discovered
+	PeersLimit = 5
+
+	// peerWeight total weight of discovered peers
+	peerWeight = 1000
+	// connectionTimeout is timeout given to connect to discovered peer
+	connectionTimeout = time.Minute
+	topic             = "full"
+	interval          = time.Second * 10
+)
+
+var (
+	// waitF calculates time to restart announcing.
+	waitF = func(ttl time.Duration) time.Duration {
+		return 7 * ttl / 8
+	}
+)
+
+// discovery implements discoverer interface to avoid using it in light availability.
+type discoverer interface {
+	findPeers(ctx context.Context)
+}
+
+// discoverer finds and stores discovered full nodes.
+type discovery struct {
+	set  *limitedSet
+	host host.Host
+	disc core.Discovery
+}
+
+// newDiscoverer constructs new Discoverer.
+func newDiscovery(h host.Host, d core.Discovery) *discovery {
+	return &discovery{
+		newLimitedSet(PeersLimit),
+		h,
+		d,
+	}
+}
+
+// handlePeersFound receives peers and tries to establish a connection with them.
+// Peer will be added to PeerCache if connection succeeds.
+func (d *discovery) handlePeersFound(topic string, peers []peer.AddrInfo) error {
+	for _, peer := range peers {
```

Comment:
Let's handle each peer in a separate routine. The Connect call inside is blocking and may take up to 1 minute.

Reply:
handlePeerFound is called in a separate routine for each peer.